contracts/contracts/strategies/GeneralizedUniswapV3Strategy.sol

```diff
@@ -0,0 +1,850 @@
+     * @dev Must be a position minted by this contract
+     * @param tokenId ERC721 token ID of the position to liquidate
+     */
+    function closePosition(uint256 tokenId)
+        external
+        onlyGovernorOrStrategistOrOperator
+        nonReentrant
+    {
+        require(tokenIdToPosition[tokenId].exists, "Invalid position");
+        _closePosition(tokenId);
+
+        // Deposit all dust back to reserve strategies
+        _depositAll();
+    }
+
+    /**
+     * @notice Closes the position denoted by the tokenId and and collects all fees
+     * @param tokenId ERC721 token ID of the position to liquidate
+     * @param amount0 Amount of token0 received after removing liquidity
+     * @param amount1 Amount of token1 received after removing liquidity
+     */
+    function _closePosition(uint256 tokenId)
+        internal
+        returns (uint256 amount0, uint256 amount1)
+    {
+        Position storage p = tokenIdToPosition[tokenId];
+
+        if (p.liquidity == 0) {
+            return (0, 0);
+        }
+
+        // Remove all liquidity
+        (amount0, amount1) = _decreaseLiquidityForPosition(p, p.liquidity);
+
+        // Collect all fees for position
+        (uint256 amount0Fee, uint256 amount1Fee) = _collectFeesForToken(
+            tokenId
+        );
+
+        amount0 = amount0 + amount0Fee;
+        amount1 = amount1 + amount1Fee;
+
+        if (tokenId == currentPositionTokenId) {
+            currentPositionTokenId = 0;
+        }
+
+        emit UniswapV3PositionClosed(tokenId, amount0, amount1);
+    }
+
+    /**
+     * @notice Mints a new position on the pool and provides liquidity to it
+     * @param desiredAmount0 Desired amount of token0 to provide liquidity
+     * @param desiredAmount1 Desired amount of token1 to provide liquidity
+     * @param lowerTick Lower tick index
+     * @param upperTick Upper tick index
+     * @return tokenId ERC721 token ID of the position minted
+     * @return liquidity Amount of liquidity added to the pool
+     * @return amount0 Amount of token0 added to the position
+     * @return amount1 Amount of token1 added to the position
+     */
+    function _mintPosition(
+        uint256 desiredAmount0,
+        uint256 desiredAmount1,
+        int24 lowerTick,
+        int24 upperTick
+    )
+        internal
+        returns (
+            uint256 tokenId,
+            uint128 liquidity,
+            uint256 amount0,
+            uint256 amount1
+        )
+    {
+        INonfungiblePositionManager.MintParams memory params = INonfungiblePositionManager
+            .MintParams({
+                token0: token0,
+                token1: token1,
+                fee: poolFee,
+                tickLower: lowerTick,
+                tickUpper: upperTick,
+                amount0Desired: desiredAmount0,
+                amount1Desired: desiredAmount1,
+                amount0Min: maxSlippage == 0
```

Comment:
was thinking of the same thing. The operator should have a clear view of the state of Uniswap pool without the dangers of front running bots altering it just before the transaction executes. It would make sense to pass the min amounts as parameters to the external function